Doc/reference/expressions.rst

```diff
@@ -1687,6 +1687,11 @@ Or, when processing a file stream in chunks:
    while chunk := file.read(9000):
        process(chunk)
 
+Assignment expressions must be surrounded by parentheses when used
+as sub-expressions in subscript, slicing, conditional, lambda,
+keyword-argument, generator, and comprehension-if expressions.  They
+can only be used as is in if and while statements and in decorators.
```

Comment:
" and in decorators" must be removed in the 3.8 backport.

@gvanrossum Have there been any other changes that you know of?

Reply:
I don't think so. The plan was to look for other places where the parentheses could be omitted without confusing either humans or parsers, but we never did that, and I don't think there are any pressing needs.

Reply:
PEP-572 mentioned the extreme possibilities of requiring parentheses always and only where the parser required them, with the first rejected because if and while uses, as in the doc examples, were thought to be most common.  Most of the sub-expressions requiring parentheses were pretty clearly better with them after I tried without.  The most puzzling is why [a:= i+j for j in range(3)] is legal while  (a:= i+j for j in range(3)) is not.  (Since only the last value gets preserved, this seems like a relatively rare use.)

I recently used a.es for debugging (a use not mentioned in the PEP), to pull multiple values out a complex expression for a single debug print.  Positional arguments were easy; indexes required ()s.  The latter is probably the first extension of () not required I would want looked at.

Reply:
Yeah, it makes sense for `a[i := f()]` and for `a[i:= f(), j := g()]` etc., though I would prefer to require the parentheses in slices (too many colons otherwise :-). Provided we can exclude it in slices, it should be a simple change to the grammar to support it in indexes. Maybe @lysnikolaou can help?

Reply:
Yup, that's a very simple change to the parser. I'll have a PR ready later tonight.